### ISSUE TITLE
Additional Remote Functionality for Kaleidescape

### ISSUE THREAD

**Issue by thrillcat** [SteveEasley/pykaleidescape#3]
Would it be possible to add more remote functionality, such as adding the intermission button, the info button, play/pause, etc?

I’m trying to eliminate as much IR as possible in my system. I’m using a lot of this integration but have also had to learn several IR commands from the remote itself to round out the controls in my Unfolded Circle remote.

**Comment by danteoh** [comment]
@thrillcat I think my PR helps solve this - it allows unregistered events to be observed, so that in the future you don't need code changes to this library when the KScape event list expands.

**Comment by jgoldman1** [comment]
@danteoh  & @SteveEasley Does your answer above handle the case for detecting events "USER_DEFINED_EVENT:VOLUME_UP:" and "USER_DEFINED_EVENT:VOLUME_DOWN:" event that get sent by the newest players when the native Kaleidescape Remote Control (gen 2) have their volume up & down pressed?

Trying to detect those messages from the player and then make an audio system raise or lower the volume.

Thanks!

**Comment by danteoh** [comment]
@jgoldman1 Yes, that's exactly why I did it - I have a Strato V and wanted the ability to have it control my denon without waiting for Kaleidescape to ship the HDMI volume control functionality which has no ETA.

I'm assuming you're using Home Assistant - the next step after this is to wire these update into HA Core - at that point it has to broadcast the "volume capabilities" event to the KScape unit (I had it in this core repo but decided that's probably not the best place to do it).  I haven't heard from @SteveEasley  yet - not sure if he's maintaining this anymore.

**Comment by jgoldman1** [comment]
Thanks, that's great and my exact use case... yes, i have a big HA install and use it with Kaleidescape for lighting triggers.

Any idea how i can make these events raise and lower volume on my (HA-connected) Marantz controller?
Here's the word from a Kaleidescape employee on how it works in the player:
https://www.kscapeowners.com/XF/threads/strato-remote-volume-controls.7463/#post-79065

Thanks for any help/advice!

--josh

**Comment by danteoh** [comment]
Gimme a few days. I forked HA Core - now time to get them to accept making
my fork of pykaleidescape the mainline.

Dan

**Comment by SteveEasley** [comment]
Hey sorry for being away so long. Started developing some new features this weekend (unrelated to this request) and noticed I totally missed this.

danteoh's PR will help with volume events, but I don't think handling unregistered events will help detect `intermission` and `info` presses since I don't see any events emitted for those at all. Though for intermission its easy to detect from existing events like MOVIE_LOCATION=intermission events, and play/pause events from PLAY_STATUS change events. All of these are already exposed in HA sensors.

**Comment by kevinlester** [comment]
@jgoldman1 See my repo [here](https://github.com/kevinlester/kaleidescape_volume) for enabling a Kaleidescape with volume buttons to be able to publish volume events in HA that can be used for automations.  It's working great on my HA system to have my Kaleidescape remote be able to control the volume on my Denon receiver.  That repo is based on an updated fork of this repo that I've submitted for inclusion [here](https://github.com/SteveEasley/pykaleidescape/pull/8).  Ideally that PR can be merged and included into HA core so that everyone can enjoy that functionality without needing a separate repo.  Until then, you can use the linked repo to enable volume control.

**Comment by SteveEasley** [comment]
This feature is now fully developed and a PR created on the HA side to integrate it! https://github.com/home-assistant/core/pull/158756. Feel free to test and review!

**Comment by SteveEasley** [comment]
The HA PR over there seems to have stalled for the last 4 weeks. Might help if someone else pings it just to show interest.

**Comment by jgoldman1** [comment]
Submitted something there. Not sure if my wording was right or helpful. Let
me know what to post in what I said wasn’t right.

**Comment by SteveEasley** [comment]
Nope perfect thanks!

**Comment by kevinlester** [comment]
Same, just left a comment.  🤞

**Comment by jgoldman1** [comment]
It works!! So cool.  Thanks to @kevinlester and @SteveEasley!!
Really awesome to have this enabled on the K remote.

**Comment by jgoldman1** [comment]
Question:  I adjusted the delay to 0.1.  Lets me move the volume at a speed I prefer. (thanks for enabling that adjustability).
But now a quick "click" on the volume+ or volume- moves the volume on my Marantz by 1.0 or 1.5 steps rather than the perfect 0.5 steps that it was doing when i left delay to your default 0.25.

Is it possible maybe in a future rev of this bridge to set the "click" volume movement to say 0.5 and set the "hold" volume adjustment speed to something else?  Or do the right things not get exposed to you to handle click differently than "hold"?

**Comment by SteveEasley** [comment]
@jgoldman1 Honestly I am a little confused atm that changing delay (entity.py::DEBOUNCE_TIME right?) is affecting the volume+ and - buttons. Debounce is just related to the K mobile app's volume slider. Just for context:

1. You mention a "default 0.25" delay. Where do you see that delay? I am not aware of that specific value anywhere.
2. Are you running this in your own `custom_components` directory?
3. Where are you clicking the volume + and - buttons? K remote, K mobile app?

**Comment by jgoldman1** [comment]
OK, i'm just not technical enough to answer all your questions.  Not a dev.  just trying to work through this stuff.
Answer to your Q1:
I found the reference to 0.25 as the default delay value here: https://github.com/kevinlester/kaleidescape_volume


Q2: I have no idea what this means, so i suppose my answer would be "no"
Q3: I'm doing this on the K remote.

Sorry I'm not technical enough to give you more details.  I'm probably in over my head, but i do have the volume control working, albeit with a few quirks which is why i posted what i did with my results of setting that interval value to what i tried.  tried several values... lower does make the volume move vaster, but seems to also mean that a quick press no longer moves it on my Marantz by the 0.5 amount that it does with the delay set to 0.25.

--josh

**Comment by SteveEasley** [comment]
Oh that github custom integration is not mine (its kevinlester's). But once my new features are added it would do a similar thing. If I get some time in the next few days I will add my own HACS version of it you could try instead. Its looking like it wont be until the march HA release my changes get approved for core (not sure why its taking so long).

**Comment by jgoldman1** [comment]
> Oh that github custom integration is not mine (its kevinlester's). But once my new features are added it would do a similar thing. If I get some time in the next few days I will add my own HACS version of it you could try instead. Its looking like it wont be until the march HA release my changes get approved for core (not sure why its taking so long).

Thanks @SteveEasley and sorry for the confusion... my fault, posting in the wrong area and to the wrong person.  github makes my brain hurt.  I'll go back to my role as a regular consumer hobbyist!

**Comment by kevinlester** [comment]
@jgoldman1 - feel free to open a Github issue for any suggestions on the volume integration [here](https://github.com/kevinlester/kaleidescape_volume/issues) and I'll take a look.  As @SteveEasley mentioned, though, my repo should be short lived, since it shouldn't be necessary after the current version of this repo gets integrated into HA Core.  But until it does, i'm happy tweaking my repo to bridge the gap.

**Comment by jgoldman1** [comment]
Checking in to see the latest with whether or not the modified Kaleidescape integration was ever integrated into HA Core.  Is @kevinlester's volume-control repo still needed to get volume buttons on the K remote to control volume on a different device?

Thanks to you all for this work...

**Comment by danteoh** [comment]
@jgoldman1 kscape now supports CEC so I personally don't see a need for this anymore.

**Comment by jgoldman1** [comment]
Only on strato V, E, M. 
But I understand having other priorities
